### PR TITLE
Bigdata

### PR DESCRIPTION
**Pull request by blemoine**
Support for BigData API, high and low level.

### CODE REVIEW

**Comment by jtheoof** [comment]
The build fails on MR because the env variables are encrypted only on `mnubo` repo. I think there is a way to prevent travis CI to run on MR but I don't know how.

**Comment by jtheoof** [comment]
Minor: angular commit id is `chore` not `chores`

File: src/bigdata/Bigdata.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+const basePath = '/api/v3/bigdata';
+
+const mergeStreamResponse = (oldResponse: StreamPageResponse, newResponse: StreamPageResponse): StreamPageResponse => {
+  return Object.assign({}, newResponse, { rows: oldResponse.rows.concat(newResponse.rows) })
+};
+
+export class Bigdata {
+
```
Comment (jtheoof) [inline src/bigdata/Bigdata.ts]: minor: no new line after `class` def

File: src/bigdata/Bigdata.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import { authenticate } from '../decorators';
```
Comment (jtheoof) [inline src/bigdata/Bigdata.ts]: For consistency file name should be `bigdata.ts`

**Comment by blemoine** [comment]
Commit messages fixed

File: src/http/http.ts [hunk 3]
```diff
@@ -1,32 +1,32 @@
-import {Request, RequestMethods, RequestOptions} from './request';
-import {nodeHttpRequest} from './backends/node_backend';
+import { Request, RequestMethods, RequestOptions } from './request';
+import { nodeHttpRequest } from './backends/node_backend';
```
Comment (jtheoof) [inline src/http/http.ts]: inconsistency with the rest of the project

File: src/http/http.ts [hunk 4]
```diff
@@ -1,32 +1,32 @@
...
 
-export class http {
-  static request(request: Request): Promise<any> {
-      return nodeHttpRequest(request);
-  }
+export const http = {
```
Comment (jtheoof) [inline src/http/http.ts]: I am not sure I understand why this was changed.

File: src/bigdata/Bigdata.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+  }
+
+  @authenticate
+  startExport(query: object): Promise<StartExportResponse> {
+    return this.client.post(basePath + '/startexport', query)
+  }
+
+  @authenticate
+  streampage(pageId: string): Promise<StreamPageResponse> {
```
Comment (jtheoof) [inline src/bigdata/Bigdata.ts]: `streamPage`

File: src/bigdata/Bigdata.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
+import { authenticate } from '../decorators';
+import { Client } from '../mnubo';
```
Comment (jtheoof) [inline src/bigdata/Bigdata.ts]: inconsistency with the rest of the project

File: spec/bigdata/bigdata.spec.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+                expect(response.streamsFirstPages.length).toBeGreaterThan(0);
+                done();
+            }).catch((error) => {
+                fail(error);
+                done();
+            });
+        });
+    });
+
+    describe('.streampage()', function() {
```
Comment (jtheoof) [inline spec/bigdata/bigdata.spec.js]: `.streamPage()`

File: spec/bigdata/bigdata.spec.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
+const mnubo = require('../../dist/mnubo');
+
+describe('bigdata', function() {
+    const client = new mnubo.Client({
+        id: process.env.MNUBO_CLIENT_ID,
+        secret: process.env.MNUBO_CLIENT_SECRET,
+        env: 'sandbox'
+    });
+
+    describe('.startexport()', function() {
```
Comment (jtheoof) [inline spec/bigdata/bigdata.spec.js]: .startExport()

**Comment by daddykotex** [comment]
In fact there is a way to fix this, travis expose environment variables to let you know if the the current build is from a MR or from the main repo.

we should avoid running test when this environment variable is set.

File: src/http/http.ts (hunk 4 above)
Comment (blemoine) [inline src/http/http.ts]: It was a class with only static methods, so this in term of usage equivalent of the above (an object with functions). 
In term of implementation, "static is bad", and I think the intent is more clear without the `class`.

File: src/bigdata/bigdata.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    });
+  }
+
+  @authenticate
+  private streamToTheLastpage(pageId: string): Promise<StreamPageResponse> {
+    return this.streamPage(pageId).then(response => {
+      const nextPage = response.nextPage;
+
+      if (nextPage) {
+        return this.streamToTheLastpage(nextPage)
```
Comment (jtheoof) [inline src/bigdata/bigdata.ts]: This recursion could be an issue if there a gazillion pages. But could it really be the case? I'll double check with @daddykotex

File: src/bigdata/bigdata.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+const basePath = '/api/v3/bigdata';
+
+const mergeStreamResponse = (oldResponse: StreamPageResponse, newResponse: StreamPageResponse): StreamPageResponse => {
+  return Object.assign({}, newResponse, { rows: oldResponse.rows.concat(newResponse.rows) })
+};
+
+export class Bigdata {
+
```
Comment (jtheoof) [inline src/bigdata/bigdata.ts]: minor: can you remove this new line?

File: src/bigdata/bigdata.ts (hunk 9 above)
Comment (blemoine) [inline src/bigdata/bigdata.ts]: Ok, but I'm not sure I have any choice: the pageId for the next call is in the response of the promise.
I can do some horrible thing with mutable array. But... well... _mutation_

Comment (daddykotex) [inline src/bigdata/bigdata.ts]: no issue in sight

**Pull request merged**